src/Process/Pool/Logger.php

```diff
@@ -49,7 +49,13 @@ public function log($level, $message, array $context = [])
                 $logMessage->setProcessId($processId);
                 $logMessage->setProcessPath($this->_getProcess()->getPath());
                 $logMessage->setMessage($message);
-                fwrite(STDOUT, $this->getLogFormatter()->getFormattedMessage($logMessage) . "\n");
+                if (count($context) === 1) {
+                    $contextValue = array_pop($context);
+                    if (is_object($contextValue) && $contextValue instanceof \JsonSerializable) {
```

Comment:
a bit of a non-sequitur, but this reminds me of something important. forgive me if i'm being super pedantic here, i'm walking in blind and need to figure out where we are in our logging plans, but i want to get this out somewhere so i don't forget to bring it up:

we should start planning ahead for logging structured data to elasticsearch. right now the structured aspect of our logs is limited to metadata provided by kubernetes etc. as we move toward making the log record itself a structured data format for elasticsearch ingestion, it will be important to constrain the cardinality of arbitrary fields users can add to the context array. elasticsearch has a default limit on inverted indices of 1000 unique keys, and if you provide ES with structured data with lots of deeply nested fields, it will consider each full path in the structured data its own index (eg., `foo.bar` is a single search index, and `foo.bar.baz` is its own single search index)

because of this, limiting the number of different keys it is possible to log will become important to make sure any structured log data we index to ES is efficiently searchable, and doesn't cause our ES cluster to have issues keeping everything in memory. some of this can be dealt with my defining explicit mappings to prevent too many dynamic fields from being indexed in the first place, or by making some structured data unsearchable, but still visible in kibana.

ps. by "structured logging", i mean breaking the `log` key of a [log event like this one](https://logs.neighborhoods.com/app/kibana#/doc/63c8d5b0-f72f-11e8-b428-736910a52fcd/logstash-2018.12.21/fluentd?id=x0ma0WcBUivnU-V1eCVw&_g=()) into structured data, so that someone could search for log events by keys like `time`, `level`, `message`, etc. in addition, this particular message is including data about what looks to be a pid and a child pid. those two values could also be extracted into their own keys in a structured log event, so that an ES user could find all log events related to a specific pid, etc.

structured logging like this can be really powerful for understanding and debugging a system.

Reply:
@mucha55 - also
> (because creating all those models is slow)

prove that though

Reply:
@rhift 	 my comment above has nothing to do with this PR, this convo just got me thinking about some things. will make sure i find the correct venue to explore structured logging stuff and our current plans with our app logging, i'm totally unaware of our current plans, don't mean to add noise to your work!

Reply:
also -
> I'd rather they use primitive arrays over anonymous classes

why? I don't have any opinions about US except that this thing is serializable to JSON and that the contract sets US up to succeed wrt to this.

Reply:
> > (because creating all those models is slow)

> prove that though

I mean in terms of developer time. Even if you're writing/modifying a `OneParticularTypeOfErrorMessage.prefab.definition.yml` and fabbing it, it's still faster to use an `array` for iterating, and then fab a proper model once you settle on a schema (and we can enforce that in PRs/sonarqube)

Reply:
> > I'd rather they use primitive arrays over anonymous classes

> why? I don't have any opinions about US except that this thing is serializable to JSON and that the contract sets US up to succeed wrt to this.

By "they" I mean all users, which includes US

Reply:
US - User Space (fwiw)

Reply:
I thought about this, but it does remove type safety from the setters because (unless I'm mistaken) there's no PHP interface that includes all things that can be json_encoded, so get/set context would need to have their types removed.

I'm curious if it would be better to use an array as the type and when you want to pass a \JsonSerializable object into it, you just call `$object->jsonSerialize()`. This is a minimal addition to passing in a serializable object compared to declaring an anonymous `\JsonSerializable` class when you have data that isn't already in an object.

Reply:
👍 it might be interesting down the road to have different logger objects for different use-cases, so that type safety and the kinds of contextual data provided could be specified per use case. i would lean toward as specific an interface as possible to hopefully constrain log messages to the most limited vocabulary possible, makes parsing logs easy when everyone is using the same key for the same context data